mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,496 @@
+        if let Some(client_info) = this.clients.get_mut(&client_id) {
+            client_info.add_streamer(streamer_id.clone());
+            this.streamers
+                .entry(streamer_id.clone())
+                .or_insert(streamer_info)
+                .add_client(client_id);
+        } else {
+            // The client was removed while we were spawning the streamer.
+            // We no longer have a connection for it.
+            return Err(StreamingManagerError::UnknownClient);
+        }
+        Ok(streamer_id)
+    }
+
+    /// Sends data to a streamer with `streamer_id`.
+    pub fn send<T: Send + 'static>(&self, streamer_id: &str, data: T) -> Result<(), StreamingManagerError> {
+        let this = self.0.read();
+        let streamer_info = this
+            .streamers
+            .get(streamer_id)
+            .ok_or(StreamingManagerError::StreamerNotFound)?;
+        let data_in = streamer_info.data_in.as_ref().ok_or(StreamingManagerError::NoDataIn)?;
+        data_in
+            .unbounded_send(Box::new(data))
+            .map_err(|e| StreamingManagerError::SendError(e.to_string()))
+    }
+
+    /// Same as `StreamingManager::send`, but computes that data to send to a streamer using a closure,
+    /// thus avoiding computations & cloning if the intended streamer isn't running (more like the
+    /// laziness of `*_or_else()` functions).
+    ///
+    /// `data_fn` will only be evaluated if the streamer is found and accepts an input.
+    pub fn send_fn<T: Send + 'static>(
+        &self,
+        streamer_id: &str,
+        data_fn: impl FnOnce() -> T,
+    ) -> Result<(), StreamingManagerError> {
+        let this = self.0.read();
+        let streamer_info = this
+            .streamers
+            .get(streamer_id)
+            .ok_or(StreamingManagerError::StreamerNotFound)?;
+        let data_in = streamer_info.data_in.as_ref().ok_or(StreamingManagerError::NoDataIn)?;
+        data_in
+            .unbounded_send(Box::new(data_fn()))
+            .map_err(|e| StreamingManagerError::SendError(e.to_string()))
+    }
+
+    /// Stops streaming from the streamer with `streamer_id` to the client with `client_id`.
+    pub fn stop(&self, client_id: u64, streamer_id: &str) -> Result<(), StreamingManagerError> {
+        let mut this = self.0.write();
+        if let Some(client_info) = this.clients.get_mut(&client_id) {
+            client_info.remove_streamer(streamer_id);
+
+            this.streamers
+                .get_mut(streamer_id)
+                .ok_or(StreamingManagerError::StreamerNotFound)?
+                .remove_client(&client_id);
+
+            // If there are no more listening clients, terminate the streamer.
+            if this.streamers.get(streamer_id).map(|info| info.clients.len()) == Some(0) {
+                this.streamers.remove(streamer_id);
+            }
+            Ok(())
+        } else {
+            Err(StreamingManagerError::UnknownClient)
+        }
```

Comment:
Early returning error is always better as it avoids nesting and maker code more readable
```rust
        let client_info = this
            .clients
            .get_mut(&client_id)
            .ok_or(StreamingManagerError::UnknownClient)?;
        client_info.remove_streamer(streamer_id);

        this.streamers
            .get_mut(streamer_id)
            .ok_or(StreamingManagerError::StreamerNotFound)?
            .remove_client(&client_id);

        // If there are no more listening clients, terminate the streamer.
        if this.streamers.get(streamer_id).map(|info| info.clients.len()) == Some(0) {
            this.streamers.remove(streamer_id);
        }

        Ok(())
```

Reply:
Covered in [e4c778d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2172/commits/e4c778d9e0bf3545ce95f214b9dcdda29efb343a)